plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,239 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Authors: Răzvan Vîrtan <virtanrazvan@gmail.com>
+ *          Justin He     <justin.he@arm.com>
+ *
+ * Copyright (c) 2021, Arm Ltd. All rights reserved.
+ * Copyright (c) 2021, University Politehnica of Bucharest. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <libfdt.h>
+#include <kvm/config.h>
+#include <kvm-arm/smp.h>
+#include <kvm/intctrl.h>
+#include <arm/cpu.h>
+#include <kvm-arm/smp_defs.h>
+#include <uk/plat/io.h>
+#include <uk/plat/lcpu.h>
+#include <arm/psci.h>
+#include <ofw/fdt.h>
+#include <uk/arch/spinlock.h>
+
+const size_t arm64_cpu_size = sizeof(struct arm64_cpu);
+
+struct arm64_cpu cpus[CONFIG_UKPLAT_LCPU_MAXCOUNT];
+
+static uint64_t bsp_cpu_id;
+static unsigned int smp_numcores;
+void _lcpu_start(struct arm64_cpu *cpu);
```

Comment:
This was meant to keep consistency with the x86 SMP implementation
https://github.com/unikraft/unikraft/pull/244/files#diff-dedfe4391b8bffb941dc93ca511584df5949f52d1af599ac52ada6d7a18e3724R5

Reply:
I will remove the front underscores on `_lcpu_start` and `_lcpu_entry_default`. However, changing the name for `__lcpuid` should be consistent with the x86 PR, since it is a common data type.

Reply:
> As discussed, we don't need to be consistent at the implementation level. Especially if they involve questionable habits 😄 In general I would like some justification of the underscore pattern, and the underlying semantics. From what I see it is used in various places (including kvm/arm) with various semantics (eg to indicate a function written in assembly). But since these are supposed to be reserved, I would prefer either some justification or to avoid using them altogether.

Right, especially considering that the x86 implementation is not final, yet. But I like that @razvanvirtan tried to make the code as similar as possible because it makes it easy to maintain and comprehend if you are familiar with another architecture's implementation. So I am generally a big fan of "look-alike" code and implementations. But exactly as you said, we then have to choose the best implementation/habits as example.